shiny/ui/_sidebar.py

```diff
@@ -40,6 +43,68 @@
     "panel_main",
 )
 
+SidebarOpenValues = Literal["desktop", "open", "closed", "always"]
+"""
+The possible values for the `open` parameter in :func:`~shiny.ui.sidebar`.
+
+* `"desktop"`: the sidebar starts open on desktop screen, closed on mobile
+* `"open"`: the sidebar starts open
+* `"closed"`: the sidebar starts closed
+* `"always"`: the sidebar is always open and cannot be closed
+"""
+
+
+@dataclass
+class SidebarOpen:
```

Comment:
Could all of this be simplified to:

```py
SidebarOpenValue = Literal["open", "closed", "always"]

class SidebarOpenSpec(TypedDict):
    desktop: SidebarOpenValue
    mobile: SidebarOpenValue
```

Then the definition of `sidebar()` could look like this:

```py
def sidebar(
    *,
    open: Optional[Literal["desktop"] | SidebarOpenValue | SidebarOpenSpec] = None,
):
    ...
```

And here are some valid ways of calling it:

```py
sidebar(open="desktop")

sidebar(open="always")

sidebar(open={"desktop": "closed", "mobile": "always"})
```


And it might be useful to add a utility function to map from a single string, like `"open"`, to a `SidebarOpenSpec` object like, `{"desktop": "open", "mobile": "open"}`.

Reply:
I went into this PR anticipating using a `TypedDict` and then in pairing with @schloerke we landed on using a `dataclass`. I can see advantages and disadvantages to both, and I'd love to get @schloerke's opinion on this too.

To clarify, users are not _required_ to use `SidebarOpen()` to customize the sidebar open behavior, they only need it to have separate decisions on mobile and desktop.

```python
ui.sidebar(open="desktop")

ui.sidebar(open="always")

ui.sidebar(open=ui.SidebarOpen("closed", "always"))
```

Reply:
> if someone wants to customize the sidebar open behavior, they would use the `SidebarOpen` constructor

Correct

----------------------

I'm ok with TypeDict.

I had suggested a dataclass as I wanted to leave the door open for more combinations down the road. I find it easier when looking at the code that a class is being used. I can hover on it and get documentation. With a `TypedDict` definition, the dict I send in just happens to match the shape and is difficult to debug. This area of the code felt like you'd rarely use a custom value, and if you did, you'd want documentation from the constructor.

... but the custom dict being sent isn't too big and I don't know of more landscapes than mobile / desktop that we'll support in the future.

Reply:
One more thought about discoverability: if the docstring gives a couple of examples of possible values, I think that will help a lot (that would be much better than simply pointing people to the docs for the `SidebarOpenSpec` typed dict).

Reply:
I went with something in between that uses all of the above suggestions in 3cb4fd1c15a7ab4e2b9d808066f4af04acfbc751:

1. I updated `SidebarOpenValue` to be only `Literal["open", "closed", "always"]` as suggested. We still accept `open="desktop"`, but it's only referenced in the type hints and not the documentation.

2. I added `SidebarOpenSpec`, a `TypedDict` type so that `open` now accepts a `SidebarOpenValue` or a dictionary conforming to `SidebarOpenSpec` (or `"desktop"`). So users are expected to provide a string or a dictionary.

3. I kept the `SidebarOpen` dataclass, but it's now only used internally. This lets us keep the data structure and related functions together in one class, including the method that maps from a single value to the corresponding `SidebarOpenSpec`. I also added a validation method that catches unexpected values.

4. Finally, I included an example dictionary in the `ui.sidebar()` docstring for `open`.